amm-executor/src/Spectrum/Executor/OrdersExecutor/Service.hs

```diff
@@ -0,0 +1,116 @@
+import CardanoTx.Models (TxCandidate, fullTxOutDatum, fullTxOutRef, FullTxOut)
+import Core.Throw.Combinators (throwMaybe, throwEither)
+import Control.Monad.Catch (MonadThrow)
+import Streamly.Prelude as S
+import qualified CardanoTx.Interop as Interop
+import Cardano.Api (Tx)
+import qualified RIO.List as List
+import Spectrum.Executor.PoolTracker.Data.Traced (Traced(Traced, prevTxOutRef, tracedState))
+import CardanoTx.ToPlutus (ToPlutus(toPlutus))
+import Spectrum.Executor.Data.OrderState (OrderInState(InProgressOrder, SuspendedOrder))
+import qualified Spectrum.Executor.Data.State as State
+import Ouroboros.Network.Subscription.PeerState ()
+import Prelude hiding (drop)
+
+data OrdersExecutor s m = OrdersExecutor
+  { run :: s m ()
+  }
+
+mkOrdersExecutor
+  :: forall f m s env era.
+    ( IsStream s
+    , MonadAsync m
+    , MonadUnliftIO m
+    , MonadReader env f
+    , HasType (MakeLogging f m) env
+    )
+  => Backlog m
+  -> Transactions m era
+  -> PoolResolver m
+  -> PoolActions
+  -> f (OrdersExecutor s m)
+mkOrdersExecutor backlog transactions resolver poolActions = do
+  MakeLogging{..} <- askContext
+  logging         <- forComponent "OrdersExecutor"
+  pure $ OrdersExecutor
+    { run = run' logging backlog transactions resolver poolActions
+    }
+
+run'
+  :: forall s m era. (IsStream s, MonadAsync m, MonadUnliftIO m)
+  => Logging m
+  -> Backlog m
+  -> Transactions m era
+  -> PoolResolver m
+  -> PoolActions
+  -> s m ()
+run' logging@Logging{..} backlog@Backlog{..} txs resolver poolActions =
+  S.repeatM tryAcquire & S.mapM (\case
+      Just order ->
+        infoM ("Going to execute order for pool" ++ show order) >> execute' logging backlog txs resolver poolActions order
+      Nothing    ->
+        pure ()
+    )
+
+execute' :: forall m era. (MonadUnliftIO m, MonadThrow m) => Logging m -> Backlog m -> Transactions m era -> PoolResolver m -> PoolActions -> Order -> m ()
+execute' logging backlog@Backlog{suspend, drop} txs resolver poolActions order =
+  catch (executeOrder' backlog txs resolver poolActions order) (\(err :: OrderExecErr) -> case err of
+      PriceTooHigh  -> attachLogging logging PriceTooHigh order suspend $ SuspendedOrder order
+      dropErr       -> attachLogging logging dropErr order drop $ orderId order
+    )
+
+executeOrder' :: (Monad m, MonadThrow m) => Backlog m -> Transactions m era -> PoolResolver m -> PoolActions -> Order -> m ()
+executeOrder' Backlog{checkLater} Transactions{..} PoolResolver{..} poolActions order@(OnChain _ Core.AnyOrder{..}) = do
+  mPool <- resolvePool anyOrderPoolId
+  pool@(OnChain prevPoolOut Core.Pool{poolId})  <- throwMaybe (EmptyPool anyOrderPoolId) mPool
+  (txCandidate, Predicted _ predictedPool)      <- throwEither $ runOrder pool order poolActions
+  tx    <- finalizeTx txCandidate
+  pPool <- throwMaybe (PoolNotFoundInFinalTx poolId) (extractPoolTxOut pool tx)
```

Comment:
`PoolNotFoundInFinalTx` doesn't sound like a business exception.

Reply:
I mean this shouldn't happen at all

Reply:
on one way - yes, but on another extractPoolTxOut couldn't guarantee that tx contains poolOut, so on my opinion PoolNotFoundInFinalTx is normal error here